extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/MultiContainerTaskAdapter.java

```diff
@@ -0,0 +1,151 @@
+
+    // add sidecar termination support
+    addSideCarTerminationSupport(podTemplate);
+
+    // and the init container
+    podTemplate.getSpec().getInitContainers().add(getInitContainer());
+
+    // create the job
+    return buildJob(k8sTaskId, labels, annotations, podTemplate);
+  }
+
+  @VisibleForTesting
+  private Container getInitContainer()
+  {
+    return new ContainerBuilder()
+        .withName("kubexit")
+        .withImage(config.kubexitImage)
+        .withCommand("cp", "/bin/kubexit", "/kubexit/kubexit")
+        .withVolumeMounts(new VolumeMountBuilder().withMountPath("/kubexit").withName("kubexit").build())
+        .build();
+  }
+
+  static void reJiggerArgsAndCommand(Container container, boolean primary)
+  {
+    List<String> originalCommand = container.getCommand();
+    List<String> originalArgs = container.getArgs();
+    originalCommand.addAll(originalArgs);
+    String newArgs;
+    if (primary) {
+      // for primary the command is /bin/sh -c, don't need this again, only grab args
+      newArgs = Joiner.on(" ").join(originalArgs);
+    } else {
+      newArgs = Joiner.on(" ").join(originalCommand);
+    }
+    container.setCommand(Lists.newArrayList("/bin/sh", "-c"));
+    String toExecute = "/kubexit/kubexit /bin/sh -c " + "\"" + StringEscapeUtils.escapeJava(newArgs) + "\"";
+    // we don't care about exit code of sidecar containers
+    if (!primary) {
+      toExecute += " || true";
+    }
+    container.setArgs(Collections.singletonList(toExecute));
+  }
+
+  static void addSideCarTerminationSupport(PodTemplateSpec spec)
+  {
+    Volume graveyard = new VolumeBuilder().withName("graveyard")
+                                          .withNewEmptyDir()
+                                          .withMedium("Memory")
+                                          .endEmptyDir()
+                                          .build();
+    Volume kubeExit = new VolumeBuilder().withName("kubexit")
+                                         .withNewEmptyDir()
+                                         .endEmptyDir()
+                                         .build();
+    spec.getSpec().getVolumes().add(graveyard);
+    spec.getSpec().getVolumes().add(kubeExit);
+
+    VolumeMount gMount = new VolumeMountBuilder().withMountPath("/graveyard").withName("graveyard").build();
+    VolumeMount kMount = new VolumeMountBuilder().withMountPath("/kubexit").withName("kubexit").build();
+
+
+    // get the main container
+    List<Container> containers = spec.getSpec().getContainers();
+    for (int i = 0; i < containers.size(); i++) {
+      Container container = containers.get(i);
+      container.getEnv().add(new EnvVar("KUBEXIT_NAME", container.getName(), null));
+      container.getEnv().add(new EnvVar("KUBEXIT_GRAVEYARD", "/graveyard", null));
+      container.getVolumeMounts().add(gMount);
+      container.getVolumeMounts().add(kMount);
+      if (i > 0) {
```

Comment:
is there any harm in just letting this run for a pod with no sidecars? it looks like we'll just add the kubexit init container and a couple env variables but not any KUBEXIT_DEATH_DEPS (so we ought to just exit cleanly)

then the logic can all be moved into K8sTaskAdapter

Reply:
So technically there is no `harm`.  But consider the situation where a user can't pull from dockerhub, they also have to have some cicd pipeline to push docker images for security reasons.  Now this user runs druid without any sidecars, so now they have to jump through unnecessary hoops to use this patch.  If you want to run stock without sidecars, I feel the user should not have to do extra work.  If you want to run with sidecars, then this is the best way I could come up with to solve the problem.  You are totally correct it will work with a single container, but the user will have to add dependencies for no reason, I was trying to avoid that.